Add BMI Enhancements as Requested by SCHISM Developers and A Simple Test Case

jduckerOWP opened the pull request:
### SCHISM BMI Enhancements

1. Add offshore water level time step variable as a BMI input field (ETA2_dt)
2. Add source/sinks time step variable as a BMI input field (Q_dt)
3. Include a new namelist.input variable (SCHISM_NSCRIBES) as a required input variable to allow users to control I/O scribe functionality and switch between serial and parallel BMI capabilities. 
4. Add cmake build files `SCHISM.local.bmi_serial `and `SCHISM.local.bmi_parallel` to allow user to switch between SCHISM compile options between serial and parallel BMI capabilities.
5. Simplify BMI interface code to present standard

### SCHISM Test Case Instructions

1. Clone the repository
2. cd schism/src; mkdir build; cd build
3. cmake -C ../../cmake/SCHISM.local.bmi_serial ../../src/
4. make schism_bmi_driver

8. Download and unpack the [SCHISM Lake Champlain Google Drive Test Case](https://drive.google.com/file/d/17mv0OI275CAySp4dbv-JpP3zzH2TEKvv/view?usp=sharing) 
9. cd SCHISM_Lake_Champlain_BMI_test_May_2025
10. mpirun -np 1 ../schism/src/build/bin/schism_driver namelist.input
(Execution will demonstrate simple BMI functionality, including new BMI enhancements)

PhilMiller commented on src/.CMakeLists.txt.swp:
This shouldn't have been added and committed. It looks like a `vim` temporary file. Please delete and amend the commit.

jduckerOWP replied:
Thanks for catching this vim swap file. It's been deleted.

PhilMiller commented on src/BMI/CMakeLists.txt:
```diff
@@ -1,16 +1,20 @@
 cmake_minimum_required (VERSION 3.18)
 
-set(CMAKE_Fortran_PREPROCESS ON)
+set(CMAKE_Fortran_PREPROCESS)
```
Why is this changed? I think this is why you're seeing the errors about the preprocessor.

jduckerOWP replied:
Thanks Phil! I should have seen that this was a boolean flag that needed to be set for `NGEN_ACTIVE` definition. This did indeed resolve the problem for the cmake compilation of the BMI module.

PhilMiller commented on src/BMI/bmischism.f90:
```diff
@@ -1064,7 +1101,7 @@ function schism_grid_type(this, grid, type) result (bmi_status)
        type = "unstructured"
        bmi_status = BMI_SUCCESS
-    case(2,3,4,5,6)
+    case(2,3,4,5,6,7,8)
```
Please add the new variables' grids by name, and give them `type = "scalar"`, since that's what they are.

PhilMiller commented on src/CMakeLists.txt:
```diff
@@ -125,6 +125,10 @@ if (CMAKE_C_FLAGS)
 endif (CMAKE_C_FLAGS)
 include_directories(${MPI_C_INCLUDE_PATH})
 
+set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fPIC")
+set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
+set (CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fPIC")
```
1. Please never, ever say `set(CMAKE_C_FLAGS ...)` in a `CMakeLists.txt` file, especially to append additional options - it breaks a bunch of assumptions that CMake makes. (never mind that the existing SCHISM code does this for MPI options just above - it shouldn't , and it's caused me trouble)
2. This particular behavior of building the libraries as `SHARED` should be covered by my addition of [`BUILD_SHARED_LIBS`](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html) above. The conditionals added in `Core` and `Hydro` should also not be necessary.

PhilMiller replied:
For this kind of thing in general, it's much better to use the CMake functions `target_compile_options` or `add_compile_options`.

PhilMiller commented on src/Core/CMakeLists.txt:
```diff
@@ -33,7 +33,9 @@ if(${USE_PAHM})
 endif()
 
-add_library(core ${local_extra_code}
+if(${USE_BMI})
```
As I noted on `src/CMakeLists.txt`, the changes here and in `Hydro/CMakeLists.txt` should not be necessary, and should be dropped. If that breaks this for you for some reason, we should work it out together, and not paper over it with this sort of thing.

PhilMiller commented on src/BMI/bmischism.f90:
```diff
@@ -1399,40 +1497,7 @@ function schism_get_int(this, name, dest) result (bmi_status)
-      enddo
-      dest = [bnd_ind]
-      bmi_status=BMI_SUCCESS
+    !!!!!! No integer values currently advertised fo SCHISM !!!!!!
```
Is `fo` a typo?

jduckerOWP replied:
Fixed. Thanks for catching this.

PhilMiller commented on src/BMI/bmischism.f90:
```diff
@@ -1384,6 +1452,36 @@ function schism_set_at_indices_double(this, name, inds, src) result(bmi_status)
             shum2(inds(i)) = src(i)
         enddo
         bmi_status=BMI_SUCCESS
+    case("ETA2_dt")
+        time = this%model%iths * dt
+        do i = 1, size(inds)
+            th_dt2(1) = src(i)
+            ninv=time/th_dt2(1)
+            th_time2(1,1)=real(ninv,rkind)*th_dt2(1)
+            th_time2(2,1)=th_time2(1,1)+th_dt2(1)
+        enddo
+        bmi_status=BMI_SUCCESS
+    case("Q_dt")
+        time = this%model%iths * dt
+        do i = 1, size(inds)
+            if(nsources>0) then
+                th_dt3(1) = src(i)
+                ninv=time/th_dt3(1)
+                th_time3(1,1)=dble(ninv)*th_dt3(1)
+                th_time3(2,1)=th_time3(1,1)+th_dt3(1)
+                th_dt3(3) = src(i)
+                ninv=time/th_dt3(3)
+                th_time3(1,3)=dble(ninv)*th_dt3(3)
+                th_time3(2,3)=th_time3(1,3)+th_dt3(3)
+            endif
+            if(nsinks>0) then
+                th_dt3(2) = src(i)
+                ninv=time/th_dt3(2)
+                th_time3(1,2)=dble(ninv)*th_dt3(2)
+                th_time3(2,2)=th_time3(1,2)+th_dt3(2)
+            endif
+        enddo
+        bmi_status=BMI_SUCCESS
```
These cases should probably just be treated as erroneous, since the dt values are scalars, so there are no meaningful indices to loop over, or subsets to modify selectively.

jduckerOWP replied:
Yeah, I wasn't sure here if I should even comply here with specifying the `get_inds` functionality for the `Q_dt `and `ETA2_dt` BMI scalar variables. Are you suggest here to essentially just return `BMI_FAILURE` if those scalar variables are called?

PhilMiller commented on src/BMI/bmischism.f90:
```diff
@@ -709,6 +728,12 @@ function schism_grid_x(this, grid, x) result (bmi_status)
         x(:) = xsta_bmi(:)
         bmi_status = BMI_SUCCESS
+    case(SCHISM_BMI_ETA2_TIMESTEP)
+        x(:) = -9999.d0
```
Asking for the coordinates of a scalar dt variable seems like the perfect case for giving out NaN instead of a numeric sentinel value. Honestly, I think it would be even better as just `BMI_FAILURE` and not putting anything in the result.

jduckerOWP replied:
I'm leaning towards setting `BMI_FAILURE` here to indicate to users that this is essentially invalid for scalar BMI variables. This would be sort of the same logic as well for scalar BMI variables returning `BMI_FAILURE` return codes if they're called as well for `schism_set_at_indices_double`. What do you think about this?

PhilMiller replied:
Agreed. Any routine on which it just doesn't make sense to call for these variables, we should just return `BMI_FAILURE`

PhilMiller commented on cmake/SCHISM.local.bmi_serial:
```diff
@@ -0,0 +1,67 @@
+# TVD_LIM must be one of SB, VL, MM or OS for Superbee, Van Leer, Minmod, or Osher.")
+set (TVD_LIM VL CACHE STRING "Flux limiter")
+#Turn OLDIO off to use the new scribe based I/O
+set (OLDIO ON CACHE BOOLEAN "Old nc output (each rank dumps its own data)")
+
+set (USE_ATMOS ON CACHE BOOLEAN "Coupling with atmospheric model via ESMF or BMI")
```
I have settings for `OLDIO` and `USE_ATMOS` based on `USE_BMI` in `src/CMakeLists.txt`, so setting them here should be unnecessary. Also, the settings here are backwards from what I had when we were developing the BMI code - was I building incorrectly, or is this just coming from copying a different file as a starting point?

jduckerOWP replied:
I see the logic now you've made within the `src/CMakeLists.txt` file with force setting the` OLDIO` and `USE_ATMOS` options based on the boolean flag configured for `USE_BMI`. There are potentially a few issues here however with this implementation I'll list below:

1. `OLDIO` option essentially controls the schism output formatting style for SCHISM. However, it also controls the ability to utilize the scribe parallel I/O method for SCHISM, which users may want to eventually implement for a parallelized SCHISM BMI approach. The serial SCHISM BMI implementation ONLY works for the `OLDIO` option, which is why in test case at least I need to prescribe that option. I guess the main story here is that this option really should be controlled by user specification instead of the `USE_BMI` option if that makes sense?
2. The `USE_ATMOS` option for the SCHISM community is also utilized for the SCHISM NUOPC cap within their contributions to the UFS project. For now, I am actually fine if the `USE_BMI` option forces the `USE_ATMOS` option for the SCHISM BMI approach since we also need to manually set the meteorological forcing inputs from a third-party coupler, but I just want to also mention the functionality of that option utilized for other SCHISM projects. 

Let me know if you have any follow up questions or concerns. I think the main story here I'm trying to get at basically is the `OLDIO` option would vary between a SCHISM serial and concurrent execution essentially and we may only want to direct users how to control that within the `cmake/SCHISM.local.bmi_serial` and `cmake/SCHISM.local.bmi_parallel `configuration files? Within the SCHISM model setup I've forwarded you as well, the `namelist.input` file controls that functionality within the `SCHISM_NSCRIBES` and states how users should control this variable input based on the `OLDIO` option. Does this BMI approach and functionality seem to be logical to you, or do you have a different recommendation for this approach?

PhilMiller replied:
Ah, ok, I hadn't appreciated that there would be a difference between the serial and parallel cases for SCHISM used through BMI. Having more of that stuff here makes sense then.

What I put in `src/CMakeLists.txt` doesn't force anything, it just creates defaults. I'd welcome you changing what I put there to make better sense in the usage context, since you know that better.

PhilMiller commented on src/BMI/bmischism.f90:
```diff
@@ -1,10 +1,10 @@
-  use bmif_2_0_iso
-#else
+!#ifdef NGEN_ACTIVE
+!  use bmif_2_0_iso
+!#else
   use bmif_2_0
-#endif
+!#endif
```
The commenting here should probably also get undone.

jduckerOWP commented:
@PhilMiller I've just completed all SCHISM BMI updates and revised the CMake code to reflect the new `BUILD_SHARED_LIBS` option as we've worked through together yesterday afternoon. As you look into the `bmischism.f90` and `schism_bmi_driver_test.f90` updated Fortran code, you will notice some big changes with the code base. I will summarize the changes as follows:

1. Update BMI input fields to reflect new changes to variables, which include the merging of the `t0` and `t1` variables together and including the MPI communicator variable. 
2. The entire logic of the `t0` and `t1` fields altogether really didn't make any sense. Sense the `t1` variable essentially controlled the `t0` and `t1` indices of each forcing array, that BMI logic can fully control the SCHISM forcing field updates with ease. I've updated the `schism_bmi_driver_test.f90` code base to reflect the NextGen framework workflow functionality with this new approach. I would suggest going through that driver code to get an understanding of the approach I've introduced here and would greatly appreciate any feedback or comments you may have. 
3. Update documentation within `bmischism.f90` and `schism_bmi_driver_test.f90` code bases for SCHISM developers to have a full grasp on the BMI functionality in SCHISM. 
4. Implement the first round of comments and suggestions you've had from the PR thread here so far. 

The entire code base works properly now from start to finish with the CMake compilation process. The Google drive link I've shared with you here has the Lake Champlain BMI setup compatible for the `schism_bmi_driver_test.f90` executable or the NextGen workflow script you've developed in the past as well. As you go through your evaluation, feel free to make the GitHub comments as needed and I will promptly respond. If there's a large issue at bay with the new BMI approach I've suggested, then I'll be happy as well to meet with you and discuss further. The goal here really is to lay the simplest foundation possible for a coastal model BMI integration into the NextGen framework for future development with SFINCS and/or DFlowFM as well.

PhilMiller commented on src/BMI/CMakeLists.txt:
```diff
@@ -5,12 +5,12 @@ set(CMAKE_Fortran_PREPROCESS ON)
 add_subdirectory(iso_c_fortran_bmi ${CMAKE_BINARY_DIR}/iso_c_bmi)
 
-set(BMI_SOURCES bmischism.f90 schism_model_container.f90)
+set(BMI_SOURCES bmi.f90 bmischism.f90 schism_model_container.f90)
```
I was a bit skeptical of this when I saw it, and indeed, in testing, cmake fails as follows:

```
CMake Error at BMI/CMakeLists.txt:10 (add_library):
  Cannot find source file:

    bmi.f90

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm
  .ccm .cxxm .c++m .h .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90
  .f95 .f03 .hip .ispc


CMake Error at BMI/CMakeLists.txt:10 (add_library):
  No SOURCES given to target: schism_bmi
```

Why did `bmi.f90` get added here? It should be coming from the `add_subdirectory(iso_c_fortran_bmi)` above and `target_link_libraries(... iso_c_bmi)` below.

jduckerOWP replied:
@PhilMiller the `src/BMI/bmi.f90` Fortran file and the `src/BMI/iso_c_fortran_bmi/src/bmi.f90` Fortran file refer to two separate modules that is used within the `bmischism.f90` module under the following code block:

```
#ifdef NGEN_ACTIVE
  use bmif_2_0_iso
#else
  use bmif_2_0
#endif
```

The `bmif_2_0` module reference is defined in the `src/BMI/bmi.f90` Fortran file and the `bmif_2_0_iso` module reference is defined in the `src/BMI/iso_c_fortran_bmi/src/bmi.f90` Fortran file. Therefore, that `set(BMI_SOURCES bmi.f90 bmischism.f90 schism_model_container.f90)` file inclusion is necessary on my end so the `schism_bmi_driver_test` executable can properly find the `bmif_2_0 module`. Otherwise, I get the following CMake errors below during my compilation process:
![image](https://github.com/user-attachments/assets/7f9aa4be-b1f5-4e36-ac89-ba073d134242)


Regarding your CMake error that you've posted, what CMake version did you use here? The new v4.0? I wasn't running into any of those CMake issues on the Hera cluster during my compilation process, although the CMake version is older (3.26.5). I just want to understand a bit further in our comparisons here on how we've compiled SCHISM, which appears to be sensitive to CMake versions based on your recent investigation into the code base.

PhilMiller replied:
I just chatted with Nels @hellkite500 about the `bmif_2_0_iso` stuff, and we concluded that it's kind of a relic of a development path in which he was experimenting, and then decided to take things a different direction in ngen. So, I've added the BMI reference `bmi.f90` as we're actually using, and will delete the whole `iso_c_fortran_bmi` stuff here in SCHISM as extraneous.

PhilMiller commented on src/BMI/bmischism.f90:
```diff
@@ -257,25 +266,18 @@ function schism_input_var_names(this, names) result (bmi_status)
+    input_items(7) = 'V10m'    ! 10m wind speed in northward direction (m/s)
+    input_items(8) = 'SPFH2m'   ! Specific humidity (kg/kg)
+    input_items(9) = 'RAINRATE' ! Precipitation rate (kg/m^2s)
+    input_items(10) = 'ETA2_dt' ! Scalar time step for updating water level boundaries (s)
+    input_items(11) = 'Q_dt' ! Scalar time step for updating source/sink discharge sources (precipitation + inland discharges) (s)
+    input_items(12) = 'bmi_mpi_comm_handle' ! The integer value being passed to the SCHISM BMI for an MPI communicator
```
These three are somewhat distinct from the others, in that they're only meant to be set at startup, and not in advance of each `Update` call. That said, for NextGen's purposes, they can easily enough be excluded in `ngen`, so they can stay as is.

PhilMiller commented:
OK, I haven't independently tested execution, but I'm happy with the code as it stands, and can build it easily and successfully.
